### PR TITLE
Throttle requests to mitigate RateLimitExceededExceptions

### LINKED ISSUE #1989: PyGithub should provide means to throttle API requests
Sending requests to the GitHub API too quickly will result in abuse rate limit errors. This has been discussed in several issues here (#1235, #2113, #2005, #1319, #1115, #810). I'd like to propose to wrap the Requester.__requestRaw method in a throttling method that ensures subsequent calls are at least a configured number of (float) seconds apart. This way the user of the client library can rely on calls not being too frequent.

Throttling in user code adds complexity on top of the client library and duplicates code across projects that use this client library. A good example is https://github.com/clockfort/GitHub-Backup/pull/44. Especially when iterating over paginated results, throttling becomes difficult in user code.

I have wrapped the `Requester.__requestRaw` method with the following code, which allows me to configure seconds between read (`GET`) and write (all verbs other than `GET`) requests. All requests will then be deferred by the client library if called  too quickly:

https://github.com/EnricoMi/publish-unit-test-result-action/blob/4b6277f1917325e96cadc94a1116031a0ae906bb/python/publish_unit_test_results.py#L99-L122

I am delaying retrys by a urllib `Retry` instance with a backoff factor identical to the configured delays above, which is out of scope of this issue.

### PR DESCRIPTION
**Pull request by EnricoMi**
This introduces a throttling of requests to the Github REST API (1s for writes, 0.25s for reads, configurable) to mitigate secondary rate limit errors and comply with Github's  best practices: https://docs.github.com/en/rest/guides/best-practices-for-integrators?apiVersion=2022-11-28#dealing-with-secondary-rate-limits

It defines a minimum time span between consecutive requests to the GitHub API, both caused by user code calling into the PyGithub API and called by PyGithub internally (e.g. user code iterating over a paginated list). This is achieved by hooking into the lowest possible method that is used to send all requests to GitHub API: `Requester.__requestRaw`.

For each verb the timestamp of the last request (when `__requestRaw` returned) is recorded and used before the next request to compute the time that has to be waited before the request is issued. It then calls into `time.sleep()` to respect the minimum time span.

It distinguishes between the time between any request and the time between writes. GitHub expects at least one second between writing requests (e.g. `POST`): https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits

This code is being used in [EnricoMi/publish-unit-test-result-action](https://github.com/EnricoMi/publish-unit-test-result-action/blob/b9e81bf6bfe9982d9a469f2a10e660633a0b6b85/python/publish_test_results.py#L175-L198) and runs thousands of times each day.

Here is an example log:
```
2022-10-25 07:49:05 - sleeping 0.1404111385345459s before next GitHub request
2022-10-25 07:49:05 - GitHub request: POST /repos/EnricoMi/publish-unit-test-result-action/check-runs
2022-10-25 07:49:06 - sleeping 1.9985311031341553s before next GitHub request
2022-10-25 07:49:08 - GitHub request: PATCH /repos/EnricoMi/publish-unit-test-result-action/check-runs/9087539036
```

This fixes #1989.

### CODE REVIEW

**Comment by EnricoMi** [comment]
@sfdye this fixes #1989, one of the most favoured PyGithub issues, please consider this with priority during your next stint. Thanks!

**Comment by EnricoMi** [comment]
@sfdye @s-t-e-v-e-n-k rebased and conflicts resolved, check it while it's hot!

File: tests/Framework.pyi [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# this file is needed so that tox -e lint does not complain about:
+# tests/Requester.py:63: error: Incompatible types in assignment (expression has type "float", base class "BasicTestCase" defined the type as "None")
+# tests/Requester.py:64: error: Incompatible types in assignment (expression has type "float", base class "BasicTestCase" defined the type as "None")
```
Comment (JLLeitschuh) [inline tests/Framework.pyi:3]: You can't fix this in `BasicTestCase`?

Comment (EnricoMi) [reply tests/Framework.pyi:3]: Let me try to rework this.

Comment (EnricoMi) [reply tests/Framework.pyi]: Moved into the `.py` file.

**Comment by EnricoMi** [comment]
Tested this branch in https://github.com/EnricoMi/publish-unit-test-result-action/actions/runs/5271056850/jobs/9531604524#step:5:586

```
2023-06-14 19:08:22 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action …
2023-06-14 19:08:22 +0000 - github.Requester - DEBUG - sleeping 0.15410900115966797s before next GitHub request
2023-06-14 19:08:23 +0000 - github.Requester - DEBUG - POST https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs …
2023-06-14 19:08:23 +0000 - github.Requester - DEBUG - sleeping 1.9984159469604492s before next GitHub request
2023-06-14 19:08:26 +0000 - github.Requester - DEBUG - PATCH https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs/14267722611 …
2023-06-14 19:08:26 +0000 - github.Requester - DEBUG - sleeping 1.9983429908752441s before next GitHub request
2023-06-14 19:08:30 +0000 - github.Requester - DEBUG - PATCH https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs/14267722611 …
2023-06-14 19:08:30 +0000 - github.Requester - DEBUG - sleeping 1.997204065322876s before next GitHub request
2023-06-14 19:08:33 +0000 - github.Requester - DEBUG - PATCH https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs/14267722611 …
2023-06-14 19:08:33 +0000 - github.Requester - DEBUG - sleeping 1.9984889030456543s before next GitHub request
2023-06-14 19:08:36 +0000 - github.Requester - DEBUG - PATCH https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs/14267722611 …
2023-06-14 19:08:36 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/pulls/469 …
2023-06-14 19:08:36 +0000 - github.Requester - DEBUG - sleeping 0.24811387062072754s before next GitHub request
2023-06-14 19:08:37 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/commits/b6956ada953bf773226178eaa4bd63c7c71c3a38 …
2023-06-14 19:08:37 +0000 - github.Requester - DEBUG - sleeping 0.24886202812194824s before next GitHub request
2023-06-14 19:08:37 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/commits/b6956ada953bf773226178eaa4bd63c7c71c3a38/check-runs?per_page=1 …
2023-06-14 19:08:37 +0000 - github.Requester - DEBUG - sleeping 0.24963998794555664s before next GitHub request
2023-06-14 19:08:39 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/commits/b6956ada953bf773226178eaa4bd63c7c71c3a38/check-runs?per_page=100 …
2023-06-14 19:08:39 +0000 - github.Requester - DEBUG - sleeping 0.21327781677246094s before next GitHub request
2023-06-14 19:08:41 +0000 - github.Requester - DEBUG - GET https://api.github.com/repositories/290227969/commits/b6956ada953bf773226178eaa4bd63c7c71c3a38/check-runs?per_page=100&page=2 …
2023-06-14 19:08:41 +0000 - github.Requester - DEBUG - sleeping 0.22020792961120605s before next GitHub request
2023-06-14 19:08:41 +0000 - github.Requester - DEBUG - GET https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/check-runs/14263319733/annotations?per_page=100 …
2023-06-14 19:08:41 +0000 - github.Requester - DEBUG - sleeping 0.24761605262756348s before next GitHub request
2023-06-14 19:08:42 +0000 - github.Requester - DEBUG - GET https://api.github.com/repositories/290227969/check-runs/14263319733/annotations?per_page=100&page=2 …
2023-06-14 19:08:42 +0000 - github.Requester - DEBUG - sleeping 0.24695897102355957s before next GitHub request
2023-06-14 19:08:42 +0000 - github.Requester - DEBUG - GET https://api.github.com/repositories/290227969/check-runs/14263319733/annotations?per_page=100&page=3 …
2023-06-14 19:08:42 +0000 - github.Requester - DEBUG - sleeping 0.21640610694885254s before next GitHub request
2023-06-14 19:08:43 +0000 - github.Requester - DEBUG - POST https://api.github.com/graphql …
2023-06-14 19:08:43 +0000 - github.Requester - DEBUG - sleeping 1.9967451095581055s before next GitHub request
2023-06-14 19:08:46 +0000 - github.Requester - DEBUG - POST https://api.github.com/repos/EnricoMi/publish-unit-test-result-action/issues/469/comments …
```

**Review by JLLeitschuh — APPROVED** [review]
LGTM

**Pull request merged**